### PR TITLE
feat: provide a way to "split" custom steps into pipelines

### PR DESCRIPTION
**Pull request by smira**
This allows to run multiple versions of the step in different pipelines with cron schedule and promote rules, saving/loading artifacts on the way.

The design comes from the Talos pipelines.

### CODE REVIEW

File: internal/dag/base.go [hunk 1]
```diff
@@ -29,17 +29,6 @@ func (node *BaseNode) Inputs() []Node {
 	return node.inputs
 }
 
-// InputNames implements Node interface.
```
Comment (smira) [inline internal/dag/base.go:32]: this was unused, gc'ed

File: internal/dag/node.go [hunk 2]
```diff
@@ -32,29 +33,36 @@ func Not(condition NodeCondition) NodeCondition {
 }
 
 // GatherMatchingInputNames scans all the inputs and returns those which match the condition.
-//
-// If direct input doesn't match a condition, search continues up until matching node is found.
 func GatherMatchingInputNames(node Node, condition NodeCondition) []string {
-	result := []string{}
+	return slices.Map(GatherMatchingInputs(node, condition), func(input Node) string {
```
Comment (smira) [inline internal/dag/node.go:37]: make it more DRY

File: internal/dag/node.go [hunk 3]
```diff
@@ -32,29 +33,36 @@ func Not(condition NodeCondition) NodeCondition {
...
+			result = append(result, input)
 		}
 	}
 
 	return result
 }
 
-// GatherMatchingInputs scans all the inputs and returns those which match the condition.
-//
-// If direct input doesn't match a condition, search continues up until matching node is found.
```
Comment (smira) [inline internal/dag/node.go:51]: this function was never recursive, and the comment is misleading

**Review by frezbo — APPROVED** [review]

File: internal/project/custom/custom.go [hunk 4]
```diff
@@ -137,26 +145,80 @@ func (step *Step) CompileDrone(output *drone.Output) error {
...
+			droneStep.Environment(k, v)
+		}
+
+		pipeline.Step(droneStep.DependsOn("load-artifacts"))
 	}
 
-	output.Step(droneStep)
+	if len(step.Drone.Pipelines) > 0 {
+		// add a "save artifacts" step to the default pipeline
+		output.Step(
```
Comment (frezbo) [inline internal/project/custom/custom.go:208]: do we always need it? maybe another boolean in the yaml to enabled load/save :thinking:

**Review by frezbo — APPROVED** [review]

File: internal/dag/node.go [hunk 5]
```diff
@@ -32,29 +33,36 @@ func Not(condition NodeCondition) NodeCondition {
...
 func GatherMatchingInputNames(node Node, condition NodeCondition) []string {
-	result := []string{}
+	return slices.Map(GatherMatchingInputs(node, condition), func(input Node) string {
+		return input.Name()
+	})
+}
+
+// GatherMatchingInputs scans all the inputs and returns those which match the condition.
+func GatherMatchingInputs(node Node, condition NodeCondition) []Node {
+	result := []Node{}
```
Comment (DmitriyMV) [inline internal/dag/node.go:44]: Something like `return slices.Filter(node.Inputs(), func(input Node) bool { return condition(input) })` should do the trick.

File: internal/output/drone/drone.go [hunk 6]
```diff
@@ -156,7 +223,20 @@ func (o *Output) GenerateFile(filename string, w io.Writer) error {
...
-	o.defaultPipeline.Volumes = o.volumes
+	for _, r := range o.manifest.Resources {
+		if pipeline, ok := r.(*yaml.Pipeline); ok {
+			pipeline.Volumes = o.volumes
+		}
+	}
+
+	// fix up notify pipeline
+	for _, r := range o.manifest.Resources {
+		if pipeline, ok := r.(*yaml.Pipeline); ok {
```
Comment (DmitriyMV) [inline internal/output/drone/drone.go:234]: Both `for` cycles are the same - can't we do it in one go?

File: internal/output/drone/service.go [hunk 7]
```diff
@@ -7,10 +7,16 @@ package drone
...
+func (o *Output) Service(spec *yaml.Container) {
+	o.appendService(spec, o.defaultPipeline)
+}
 
-	o.defaultPipeline.Services = append(o.defaultPipeline.Services, spec)
+func (o *Output) appendService(originalService *yaml.Container, pipeline *yaml.Pipeline) {
+	// perform a shallow copy of the step to avoid modifying the original
+	spec := *originalService
+
+	spec.Volumes = append([]*yaml.VolumeMount{}, spec.Volumes...)
```
Comment (DmitriyMV) [inline internal/output/drone/service.go:18]: We really [should](https://pkg.go.dev/golang.org/x/exp/slices#Clone) get `slices.Clone(slc)`

File: internal/project/custom/custom.go (hunk 4 above)
Comment (smira) [reply internal/project/custom/custom.go:208]: it's only if the pipelines are enabled

Comment (smira) [reply internal/project/custom/custom.go:208]: but I think without load/save it's quite useless

**Pull request merged**